### ISSUE TITLE
Can not close main window when tray is disabled

### ISSUE THREAD

**Issue by felipecrs** [hluk/CopyQ#1590]
**Describe the bug**
When the tray is disabled, the main window can not be closed once opened.

**To Reproduce**
1. Optionally reset to the default settings
2. Disable the tray under the settings
3. Open the main window
4. Try to close it (it won't)

**Expected behavior**
The main window was expected to close, as it does when the tray is enabled.

**Screenshots**
![1Bbqi5kdrX](https://user-images.githubusercontent.com/29582865/108252820-d411ac00-7137-11eb-981a-0a56fd241e8f.gif)

**Version, OS and Environment**
(Get details from `copyq version` command if possible.)
 - Application Version: 3.13.0
 - OS: Ubuntu 20.04.2
 - Desktop environment, window manager: default from Ubuntu 20.04.2 (GNOME + X11)

**Additional context**
I didn't try to reproduce in other environments.

**Comment by hluk** [comment]
This is expected - it allows users to open the window without needing to have global shortcuts configured or know the right command.

You can enable hiding window by enabling "Hide main window" option from Layout configuration tab, or with command:

```bash
copyq config hide_main_window true
```

**Comment by felipecrs** [comment]
> This is expected - it allows users to open the window without needing to have global shortcuts configured or know the right command.

I think the most intuitive approach for such cases would be to open CopyQ from the application drawer. If the user wanted it to be "pinned", one can add it to favorites. Clicking in exit but keeping the application open sounds like an issue to me.

However, that would only work if the CopyQ icon in the application drawer actually would open the main window. The exec in the `.desktop` can be tweaked to run `copyq show` rather than just `copyq`:

![x8jHpNY0FL](https://user-images.githubusercontent.com/29582865/108359238-85632100-71ce-11eb-80da-c8ed68341d98.gif)

> You can enable hiding window by enabling "Hide main window" option from Layout configuration tab, or with command:
> 
> ```shell
> copyq config hide_main_window true
> ```

It works like a charm, despite it brings one unintended side effect to me: when I deliberately open the main window, I don't want it to close when it loses focus, but with this setting it does.

So, instead of turning the "hide_main_window" option to `true`, it would be desired if CopyQ had an option like "Keep the main window opened when system tray is disabled", so I could turn it off without the side effect I just described.

To summarize, my suggestion:

1. Use `copyq show` in the desktop icon
2. Add a new option "Keep the main window opened when system tray is disabled"
3. Consider leaving it disabled by default
4. If number 3 is accepted, you may consider not including option 2 at all

**Comment by hluk** [comment]
> I think the most intuitive approach for such cases would be to open CopyQ from the application drawer. If the user wanted it to be "pinned", one can add it to favorites. Clicking in exit but keeping the application open sounds like an issue to me.

This sound like it would work only in Gnome.

> However, that would only work if the CopyQ icon in the application drawer actually would open the main window. The exec in the `.desktop` can be tweaked to run `copyq show` rather than just `copyq`

This won't work when you want to launch the app -- `copyq show` would fail because the GUI server is not running (launched with `copyq`).

Maybe it can be something like: `sh -c "copyq && copyq show"`

> > You can enable hiding window by enabling "Hide main window" option from Layout configuration tab, or with command:
> > ```shell
> > copyq config hide_main_window true
> > ```
> 
> It works like a charm, despite it brings one unintended side effect to me: when I deliberately open the main window, I don't want it to close when it loses focus, but with this setting it does.

Closing window when unfocused is controlled by another option ("Close When Unfocused" in "General" config tab):

```shell
copyq config close_on_unfocus false
```

**Comment by felipecrs** [comment]
> This sound like it would work only in Gnome.

Not only Gnome, as most of the desktop environments have pin/favorites mechanism. But even then, that's why an option to let it work how it is today would come in handy. Nonetheless, clicking in "Close" and the window not closing isn't desired at all, this is a high-level design issue that will probably not get interpreted as "intended behavior" for those unaware.

> Maybe it can be something like: sh -c "copyq && copyq show"

It works, despite the first time we open it doesn't open the main window. The following works for all the scenarios:

```shell
sh -c "(copyq &) && copyq show"
```

A few points:

1. The `copyq` command doesn't end when the server is not running yet, is it desired? I would expect something like "copyq server started successfully" and exits with `0`. That way, `sh -c "copyq && copyq show"` would work.
2. Maybe `copyq show` could be smart enough to start the server if not yet started?
3. If 2 is not desired, how about an option: `copyq show --start-server`?

> Closing window when unfocused is controlled by another option ("Close When Unfocused" in "General" config tab):
> 
> ```shell
> copyq config close_on_unfocus false
> ```

It works! Thanks so much! With those settings, now CopyQ seems to work exactly how I wanted (like Windows native clipboard manager).

**Comment by hluk** [comment]
> 1. The `copyq` command doesn't end when the server is not running yet, is it desired? I would expect something like "copyq server started successfully" and exits with `0`. That way, `sh -c "copyq && copyq show"` would work.

Yes, that's expected, `copyq` launches the application process in current shell.

> 2. Maybe `copyq show` could be smart enough to start the server if not yet started?
> 
> 3. If 2 is not desired, how about an option: `copyq show --start-server`?

I like the third option, except it would be better to create a new command for this (something like `copyq launchAndShow`) so as not to break any current behavior (`copyq show --start-server` in the latest version would open tab with name "--start-server"). The new command would be set in the desktop file (but not in the one for autostart).

**Comment by felipecrs** [comment]
Another option would be `copyq --start-server show` then.

**Comment by hluk** [comment]
> Another option would be `copyq --start-server show` then.

I think that sounds good! So it would start server in background and treat the rest of the command normally (`copyq show`). This is actually pretty handy for other stuff:

```bash
# Make sure the server is running and execute any command
copyq --start-server "copy('Some text to copy to clipboard')"

# Use separate app session for notes without leaving it running
copyq --session notes --start-server "write('NOTE'); exit()"
copyq --session notes --start-server "read(0); exit()"
```

**Comment by felipecrs** [comment]
This is awesome. One question:

In the first run, when opening from the application drawer, wouldn't the server close if I close the main window? I'll will not be an issue for me I think since I use the "Autostart" option, but just to make sure.

**Comment by hluk** [comment]
> In the first run, when opening from the application drawer, wouldn't the server close if I close the main window?

That should not happen. The server exits only in these cases:

- Select Exit from File menu (or trigger it using a shortcut).
- Getting signal from session manager, for example when logging out or shutting down the OS
- On `copyq quit` command (`quit()` from a script)
- On a signal, like `pkill copyq` or Ctrl+C from the console where you started `copyq`
- And ... well ... if something unexpected happens (a bug, X11 session stops, etc.)
